### PR TITLE
sign: bound-ify sig_alg APIs

### PR DESCRIPTION
**Pull request by woodruffw**
See #10676.

This unfortunately taints a few certificate, CRL, etc. APIs in the process. 

However, each is a singular top-level API, so the diff isn't too bad.

### CODE REVIEW

File: src/rust/src/types.rs [hunk 1]
```diff
@@ -28,6 +28,21 @@ impl LazyPyImport {
             })
             .map(|p| p.as_ref(py))
     }
+
+    pub fn get_bound<'p>(
```
Comment (alex) [inline src/rust/src/types.rs:32]: `self.get()` should just call this I think.

File: src/rust/src/x509/sign.rs [hunk 2]
```diff
@@ -442,20 +444,22 @@ fn hash_oid_py_hash(
...
+) -> CryptographyResult<pyo3::Bound<'p, pyo3::PyAny>> {
     let sig_oids_to_hash = types::SIG_OIDS_TO_HASH.get(py)?;
     match &signature_algorithm.params {
         common::AlgorithmParameters::RsaPss(opt_pss) => {
             let pss = opt_pss.as_ref().ok_or_else(|| {
                 pyo3::exceptions::PyValueError::new_err("Invalid RSA PSS parameters")
             })?;
-            hash_oid_py_hash(py, pss.hash_algorithm.oid().clone())
+            Ok(hash_oid_py_hash(py, pss.hash_algorithm.oid().clone())?
+                .to_object(py)
```
Comment (alex) [inline src/rust/src/x509/sign.rs:455]: I think you want `as_borrowed().to_owned()`, not `to_object().into_bound()`

File: src/rust/src/x509/sign.rs [hunk 3]
```diff
@@ -442,20 +444,22 @@ fn hash_oid_py_hash(
...
+                .into_bound(py))
         }
         _ => {
             let py_sig_alg_oid = oid_to_py_oid(py, signature_algorithm.oid())?;
             let hash_alg = sig_oids_to_hash.get_item(py_sig_alg_oid);
             match hash_alg {
-                Ok(data) => Ok(data),
+                Ok(data) => Ok(data.to_object(py).into_bound(py)),
```
Comment (alex) [inline src/rust/src/x509/sign.rs:462]: Same

File: src/rust/src/types.rs (hunk 1 above)
Comment (woodruffw) [reply src/rust/src/types.rs:24]: I did this via `Ok(self.get_bound(py)?.into_gil_ref())`, not sure if that's optimal :slightly_smiling_face:

**Review by alex — APPROVED** [review]

**Pull request merged**